RFC-0010-access-control-filters-masks.md

```diff
@@ -0,0 +1,204 @@
+    {
+        return Collections.emptyList();
+    }
+
+    /**
+     * Bulk method for getting column masks for a subset of columns in a table.
+     * <p>
+     * Each mask must be a scalar SQL expression of a type coercible to the type of the column being masked. The expression
+     * must be written in terms of columns in the table.
+     *
+     * @return a mapping from columns to masks, or an empty map if not applicable. The keys of the return Map are a subset of {@code columns}.
+     */
+    default Map<ColumnMetadata, ViewExpression> getColumnMasks(Identity identity, AccessControlContext context, CatalogSchemaTableName tableName, List<ColumnMetadata> columns)
+    {
+        return Collections.emptyMap();
+    }
+```
+
+ViewExpression class to hold a filter/mask expression
+
+```java
+public ViewExpression(String identity, Optional<String> catalog, Optional<String> schema, String expression)
+```
+
+Analysis.java will hold filters and masks for the table with additional methods
+
+```java
+void registerTableForRowFiltering(QualifiedObjectName table, String identity)
+
+boolean hasRowFilter(QualifiedObjectName table, String identity)
+
+void addRowFilter(Table table, Expression filter)
+
+List<Expression> getRowFilters(Table node)
+
+void registerTableForColumnMasking(QualifiedObjectName table, String column, String identity)
+
+boolean hasColumnMask(QualifiedObjectName table, String column, String identity)
+
+void addColumnMask(Table table, String column, Expression mask)
+
+Map<String, Expression> getColumnMasks(Table table)
+```
+
+#### Additional information
+
+1. What modules are involved
+   2. `presto-main`
+   3. `presto-spi`
+   4. `presto-analyzer` to hold masks and filters
+   5. `presto-hive` for legacy and sql access control
+2. Any new terminologies/concepts/SQL language additions
+   3. NA
+3. Method/class/interface contracts which you deem fit for implementation.
+   4. NA
+4. Code flow using bullet points or pseudo code as applicable
+   5. During analysis phase, access control apis used to retrieve and analyze row filters and column masks.
+   6. Analyzed filters and masks are stored in `Analysis`.
+   7. `RelationPlanner` will the get the filters and masks from `Analysis` and rewrite a new plan with them applied.
+   8. By default no filters or masks are added and the plan will not be rewritten.
+5. Any new user facing metrics that can be shown on CLI or UI.
+   1. NA
+
+## [Optional] Metrics
+
+This is a 0 to 1 feature and will not have any metrics.
+
+## [Optional] Other Approaches Considered
+
+Discussed at https://github.com/prestodb/presto/pull/21913#issuecomment-2050279419 is an approach to use the existing SPI for connector
+optimization to rewrite plan during the optimization phase. The benefits of the proposed design over this approach is that it applies
+globally to all connectors. Also, since it is an existing design that has already been in use
```

Comment:
Unfinished

Reply:
We should also probably mention benefits of conformity to the Trino SPI (ease of migration).